File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/declarations/RecordDeclaration.java
/*
 * Copyright (c) 2020, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.declarations;

import static de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge.unwrap;

import de.fraunhofer.aisec.cpg.graph.*;
import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge;
import de.fraunhofer.aisec.cpg.graph.statements.Statement;
import de.fraunhofer.aisec.cpg.graph.types.ObjectType;
import de.fraunhofer.aisec.cpg.graph.types.Type;
import de.fraunhofer.aisec.cpg.graph.types.TypeParser;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.jetbrains.annotations.NotNull;
import org.neo4j.ogm.annotation.Relationship;
import org.neo4j.ogm.annotation.Transient;

/** Represents a C++ union/struct/class or Java class */
public class RecordDeclaration extends Declaration implements DeclarationHolder, StatementHolder {

  /** The kind, i.e. struct, class, union or enum. */
  private String kind;

  @Relationship(value = "FIELDS", direction = "OUTGOING")
  @SubGraph("AST")
  private List<PropertyEdge<FieldDeclaration>> fields = new ArrayList<>();

  @Relationship(value = "METHODS", direction = "OUTGOING")
  @SubGraph("AST")
  private List<PropertyEdge<MethodDeclaration>> methods = new ArrayList<>();

  @Relationship(value = "CONSTRUCTORS", direction = "OUTGOING")
  @SubGraph("AST")
  private List<PropertyEdge<ConstructorDeclaration>> constructors = new ArrayList<>();

  @Relationship(value = "RECORDS", direction = "OUTGOING")
  @SubGraph("AST")
  private List<PropertyEdge<RecordDeclaration>> records = new ArrayList<>();

  @Relationship(value = "TEMPLATES", direction = "OUTGOING")
  @SubGraph("AST")
  private List<PropertyEdge<TemplateDeclaration>> templates = new ArrayList<>();

  /** The list of statements. */
  @Relationship(value = "STATEMENTS", direction = "OUTGOING")
  @NotNull
  private @SubGraph("AST") List<PropertyEdge<Statement>> statements = new ArrayList<>();

  @Transient private List<Type> superClasses = new ArrayList<>();
  @Transient private List<Type> implementedInterfaces = new ArrayList<>();

  @org.neo4j.ogm.annotation.Relationship
  private Set<RecordDeclaration> superTypeDeclarations = new HashSet<>();

  private List<String> importStatements = new ArrayList<>();
  @org.neo4j.ogm.annotation.Relationship private Set<Declaration> imports = new HashSet<>();
  // Methods and fields can be imported statically
  private List<String> staticImportStatements = new ArrayList<>();

  @org.neo4j.ogm.annotation.Relationship
  private Set<ValueDeclaration> staticImports = new HashSet<>();

  /**
   * It is important to set this name to a full qualified name (FQN).
   *
   * @param name the FQN
   */
  @Override
  public void setName(@NotNull String name) {
    // special case for record declarations! Constructor names need to match
    // super.setName(name);
    super.setFullName(new Name(name, null, this.getLanguage().getNamespaceDelimiter()));
    // TODO: It should be like this:
    // super.setFullName(Name.Companion.parse(name, this.getLanguage().getNamespaceDelimiter(),
    // this.getLanguage().getNameSplitter()));
    for (PropertyEdge<ConstructorDeclaration> constructorEdge : constructors) {
      constructorEdge
          .getEnd()
          .setFullName(
              new Name(
                  this.getFullName().getLocalName(),
                  this.getFullName(),
                  this.getLanguage().getNamespaceDelimiter()));
    }
  }

  public String getKind() {
    return kind;
  }

  public void setKind(String kind) {
    this.kind = kind;
  }

  public List<FieldDeclaration> getFields() {
    return unwrap(this.fields);
  }

  public List<PropertyEdge<FieldDeclaration>> getFieldsPropertyEdge() {
    return this.fields;
  }

  public void addField(FieldDeclaration fieldDeclaration) {
    addIfNotContains(this.fields, fieldDeclaration);
  }

  public void removeField(FieldDeclaration fieldDeclaration) {
    this.fields.removeIf(propertyEdge -> propertyEdge.getEnd().equals(fieldDeclaration));
  }

  public void setFields(List<FieldDeclaration> fields) {
    this.fields = PropertyEdge.transformIntoOutgoingPropertyEdgeList(fields, this);
  }

  public List<MethodDeclaration> getMethods() {
    return unwrap(this.methods);
  }

  public List<PropertyEdge<MethodDeclaration>> getMethodsPropertyEdge() {
    return this.methods;
  }

  public void addMethod(MethodDeclaration methodDeclaration) {
    addIfNotContains(this.methods, methodDeclaration);
  }

  public void removeMethod(MethodDeclaration methodDeclaration) {
    this.methods.removeIf(propertyEdge -> propertyEdge.getEnd().equals(methodDeclaration));
  }

  public void setMethods(List<MethodDeclaration> methods) {
    this.methods = PropertyEdge.transformIntoOutgoingPropertyEdgeList(methods, this);
  }

  public List<ConstructorDeclaration> getConstructors() {
    return unwrap(this.constructors);
  }

  public List<PropertyEdge<ConstructorDeclaration>> getConstructorsPropertyEdge() {
    return this.constructors;
  }

  public void setConstructors(List<ConstructorDeclaration> constructors) {
    this.constructors = PropertyEdge.transformIntoOutgoingPropertyEdgeList(constructors, this);
  }

  public void addConstructor(ConstructorDeclaration constructorDeclaration) {
    addIfNotContains(this.constructors, constructorDeclaration);
  }

  public void removeConstructor(ConstructorDeclaration constructorDeclaration) {
    this.constructors.removeIf(
        propertyEdge -> propertyEdge.getEnd().equals(constructorDeclaration));
  }

  public List<RecordDeclaration> getRecords() {
    return unwrap(this.records);
  }

  public List<PropertyEdge<RecordDeclaration>> getRecordsPropertyEdge() {
    return this.records;
  }

  public void setRecords(List<RecordDeclaration> records) {
    this.records = PropertyEdge.transformIntoOutgoingPropertyEdgeList(records, this);
  }

  public void removeRecord(RecordDeclaration recordDeclaration) {
    this.records.removeIf(propertyEdge -> propertyEdge.getEnd().equals(recordDeclaration));
  }

  public List<TemplateDeclaration> getTemplates() {
    return unwrap(this.templates);
  }

  public List<PropertyEdge<TemplateDeclaration>> getTemplatesPropertyEdge() {
    return this.templates;
  }

  public void setTemplates(List<TemplateDeclaration> templates) {
    this.templates = PropertyEdge.transformIntoOutgoingPropertyEdgeList(templates, this);
  }

  public void removeTemplate(TemplateDeclaration templateDeclaration) {
    this.templates.removeIf(propertyEdge -> propertyEdge.getEnd().equals(templateDeclaration));
  }

  @NotNull
  public List<Declaration> getDeclarations() {
    var list = new ArrayList<Declaration>();
    list.addAll(this.getFields());
    list.addAll(this.getMethods());
    list.addAll(this.getConstructors());
    list.addAll(this.getRecords());
    list.addAll(this.getTemplates());

    return list;
  }

  /**
   * Combines both implemented interfaces and extended classes. This is most commonly what you are
   * looking for when looking for method call targets etc.
   *
   * @return concatenation of {@link #getSuperClasses()} and {@link #getImplementedInterfaces()}
   */
  public List<Type> getSuperTypes() {
    return Stream.of(superClasses, implementedInterfaces)
        .flatMap(Collection::stream)
        .collect(Collectors.toList());
  }

  /**
   * The classes that are extended by this one. Usually zero or one, but in C++ this can contain
   * multiple classes
   *
   * @return extended classes
   */
  public List<Type> getSuperClasses() {
    return superClasses;
  }

  public void setSuperClasses(List<Type> superClasses) {
    this.superClasses = superClasses;
  }

  /**
   * Adds a type to the list of super classes for this record declaration.
   *
   * @param superClass the super class.
   */
  public void addSuperClass(Type superClass) {
    this.superClasses.add(superClass);
  }

  /**
   * Interfaces implemented by this class. This concept is not present in C++
   *
   * @return the list of implemented interfaces
   */
  public List<Type> getImplementedInterfaces() {
    return implementedInterfaces;
  }

  public void setImplementedInterfaces(List<Type> implementedInterfaces) {
    this.implementedInterfaces = implementedInterfaces;
  }

  public Set<RecordDeclaration> getSuperTypeDeclarations() {
    return superTypeDeclarations;
  }

  public void setSuperTypeDeclarations(Set<RecordDeclaration> superTypeDeclarations) {
    this.superTypeDeclarations = superTypeDeclarations;
  }

  public Set<Declaration> getImports() {
    return imports;
  }

  public void setImports(Set<Declaration> imports) {
    this.imports = imports;
  }

  public Set<ValueDeclaration> getStaticImports() {
    return staticImports;
  }

  public void setStaticImports(Set<ValueDeclaration> staticImports) {
    this.staticImports = staticImports;
  }

  public List<String> getImportStatements() {
    return importStatements;
  }

  public void setImportStatements(List<String> importStatements) {
    this.importStatements = importStatements;
  }

  public List<String> getStaticImportStatements() {
    return staticImportStatements;
  }

  public void setStaticImportStatements(List<String> staticImportStatements) {
    this.staticImportStatements = staticImportStatements;
  }

  @Override
  public @NotNull List<PropertyEdge<Statement>> getStatementEdges() {
    return this.statements;
  }

  @Override
  public void setStatementEdges(@NotNull List<PropertyEdge<Statement>> statements) {
    this.statements = statements;
  }

  @Override
  public String toString() {
    return new ToStringBuilder(this, Node.TO_STRING_STYLE)
        .appendSuper(super.toString())
        .append("name", getName())
        .append("kind", kind)
        .append("superTypeDeclarations", superTypeDeclarations)
        .append("fields", fields)
        .append("methods", methods)
        .append("constructors", constructors)
        .append("records", records)
        .toString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof RecordDeclaration)) {
      return false;
    }
    RecordDeclaration that = (RecordDeclaration) o;
    return super.equals(that)
        && Objects.equals(kind, that.kind)
        && Objects.equals(this.getFields(), that.getFields())
        && PropertyEdge.propertyEqualsList(fields, that.fields)
        && Objects.equals(this.getMethods(), that.getMethods())
        && PropertyEdge.propertyEqualsList(methods, that.methods)
        && Objects.equals(this.getConstructors(), that.getConstructors())
        && PropertyEdge.propertyEqualsList(constructors, that.constructors)
        && Objects.equals(this.getRecords(), that.getRecords())
        && PropertyEdge.propertyEqualsList(records, that.records)
        && Objects.equals(superClasses, that.superClasses)
        && Objects.equals(implementedInterfaces, that.implementedInterfaces)
        && Objects.equals(superTypeDeclarations, that.superTypeDeclarations);
  }

  @Override
  public int hashCode() {
    return super.hashCode();
  }

  @Override
  public void addDeclaration(@NotNull Declaration declaration) {
    if (declaration instanceof ConstructorDeclaration) {
      addIfNotContains(this.constructors, (ConstructorDeclaration) declaration);
    } else if (declaration instanceof MethodDeclaration) {
      addIfNotContains(this.methods, (MethodDeclaration) declaration);
    } else if (declaration instanceof FieldDeclaration) {
      addIfNotContains(this.fields, (FieldDeclaration) declaration);
    } else if (declaration instanceof RecordDeclaration) {
      addIfNotContains(this.records, (RecordDeclaration) declaration);
    } else if (declaration instanceof TemplateDeclaration) {
      addIfNotContains(this.templates, (TemplateDeclaration) declaration);
    }
  }

  /**
   * Returns a type represented by this record.
   *
   * @return the type
   */
  public Type toType() {
    var type = TypeParser.createFrom(this.getName(), getLanguage());

    if (type instanceof ObjectType) {
      // as a shortcut, directly set the record declaration. This will be otherwise done
      // later by a pass, but for some frontends we need this immediately, so we set
      // this here.
      ((ObjectType) type).setRecordDeclaration(this);
    }

    return type;
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/RecordDeclaration.kt
/*
 * Copyright (c) 2020, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.declarations

import de.fraunhofer.aisec.cpg.frontends.TranslationException
import de.fraunhofer.aisec.cpg.graph.*
import de.fraunhofer.aisec.cpg.graph.edges.Edge.Companion.propertyEqualsList
import de.fraunhofer.aisec.cpg.graph.edges.ast.astEdgesOf
import de.fraunhofer.aisec.cpg.graph.edges.unwrapping
import de.fraunhofer.aisec.cpg.graph.statements.Statement
import de.fraunhofer.aisec.cpg.graph.types.DeclaresType
import de.fraunhofer.aisec.cpg.graph.types.HasSecondaryTypeEdge
import de.fraunhofer.aisec.cpg.graph.types.ObjectType
import de.fraunhofer.aisec.cpg.graph.types.Type
import de.fraunhofer.aisec.cpg.persistence.DoNotPersist
import org.apache.commons.lang3.builder.ToStringBuilder
import org.neo4j.ogm.annotation.Relationship
import org.neo4j.ogm.annotation.Transient

/** Represents a C++ union/struct/class or Java class */
open class RecordDeclaration :
    Declaration(),
    DeclarationHolder,
    StatementHolder,
    EOGStarterHolder,
    DeclaresType,
    HasSecondaryTypeEdge {
    /** The kind, i.e. struct, class, union or enum. */
    var kind: String? = null

    /**
     * The [FieldDeclaration]s that are directly contained in this record declaration's AST
     * structure. This does not include any fields that might be declared in a base class or
     * interface or fields that are declared outside the AST structure.
     */
    @Relationship(value = "INNER_FIELDS", direction = Relationship.Direction.OUTGOING)
    var innerFieldEdges = astEdgesOf<FieldDeclaration>()
    /** Virtual property to directly access the nodes in [innerFieldEdges]. */
    var innerFields by unwrapping(RecordDeclaration::innerFieldEdges)

    /**
     * The [MethodDeclaration]s that are directly contained in this record declaration's AST
     * structure. This does not include any methods that might be declared in a base class or
     * interface or methods that are declared outside the AST structure.
     */
    @Relationship(value = "INNER_METHODS", direction = Relationship.Direction.OUTGOING)
    var innerMethodEdges = astEdgesOf<MethodDeclaration>()
    /** Virtual property to directly access the nodes in [innerMethods]. */
    var innerMethods by unwrapping(RecordDeclaration::innerMethodEdges)

    /**
     * The [ConstructorDeclaration]s that are directly contained in this record declaration's AST
     * structure. This does not include any constructors that might be declared in a base class or
     * interface or constructors that are declared outside the AST structure.
     */
    @Relationship(value = "CONSTRUCTORS", direction = Relationship.Direction.OUTGOING)
    var innerConstructorEdges = astEdgesOf<ConstructorDeclaration>()
    /** Virtual property to directly access the nodes in [innerConstructors]. */
    var innerConstructors by unwrapping(RecordDeclaration::innerConstructorEdges)

    /**
     * The [RecordDeclaration]s that are directly contained in this record declaration's AST
     * structure. This does not include any records that might be declared in a base class or
     * interface or records that are declared outside the AST structure.
     */
    @Relationship(value = "RECORDS", direction = Relationship.Direction.OUTGOING)
    var innerRecordEdges = astEdgesOf<RecordDeclaration>()
    /** Virtual property to directly access the nodes in [innerRecordEdges]. */
    var innerRecords by unwrapping(RecordDeclaration::innerRecordEdges)

    @Relationship(value = "TEMPLATES", direction = Relationship.Direction.OUTGOING)
    var templateEdges = astEdgesOf<TemplateDeclaration>()
    var templates by unwrapping(RecordDeclaration::templateEdges)

    /** The list of statements. */
    @Relationship(value = "STATEMENTS", direction = Relationship.Direction.OUTGOING)
    override var statementEdges = astEdgesOf<Statement>()
    override var statements by unwrapping(RecordDeclaration::statementEdges)

    @Transient var superClasses: MutableList<Type> = ArrayList()

    /**
     * Interfaces implemented by this class. This concept is not present in C++
     *
     * @return the list of implemented interfaces
     */
    @Transient var implementedInterfaces = mutableListOf<Type>()

    @Relationship var superTypeDeclarations: Set<RecordDeclaration> = HashSet()

    var importStatements: List<String> = ArrayList()

    @Relationship var imports: MutableSet<Declaration> = HashSet()

    // Methods and fields can be imported statically
    var staticImportStatements: List<String> = ArrayList()

    @Relationship var staticImports: MutableSet<ValueDeclaration> = HashSet()

    fun addField(fieldDeclaration: FieldDeclaration) {
        addIfNotContains(innerFieldEdges, fieldDeclaration)
    }

    fun removeField(fieldDeclaration: FieldDeclaration) {
        innerFieldEdges.removeIf { it.end == fieldDeclaration }
    }

    fun addMethod(methodDeclaration: MethodDeclaration) {
        addIfNotContains(innerMethodEdges, methodDeclaration)
    }

    fun removeMethod(methodDeclaration: MethodDeclaration?) {
        innerMethodEdges.removeIf { it.end == methodDeclaration }
    }

    fun addConstructor(constructorDeclaration: ConstructorDeclaration) {
        addIfNotContains(innerConstructorEdges, constructorDeclaration)
    }

    fun removeConstructor(constructorDeclaration: ConstructorDeclaration?) {
        innerConstructorEdges.removeIf { it.end == constructorDeclaration }
    }

    fun removeRecord(recordDeclaration: RecordDeclaration) {
        innerRecordEdges.removeIf { it.end == recordDeclaration }
    }

    fun removeTemplate(templateDeclaration: TemplateDeclaration?) {
        templateEdges.removeIf { it.end == templateDeclaration }
    }

    @DoNotPersist
    override val declarations: List<Declaration>
        get() {
            val list = ArrayList<Declaration>()
            list.addAll(innerFields)
            list.addAll(innerMethods)
            list.addAll(innerConstructors)
            list.addAll(innerRecords)
            list.addAll(templates)
            return list
        }

    val superTypes: List<Type>
        /**
         * Combines both implemented interfaces and extended classes. This is most commonly what you
         * are looking for when looking for method call targets etc.
         *
         * @return concatenation of [.getSuperClasses] and [.getImplementedInterfaces]
         */
        get() = superClasses + implementedInterfaces

    /**
     * Adds a type to the list of super classes for this record declaration.
     *
     * @param superClass the super class.
     */
    fun addSuperClass(superClass: Type) {
        superClasses.add(superClass)
    }

    override fun toString(): String {
        return ToStringBuilder(this, TO_STRING_STYLE)
            .appendSuper(super.toString())
            .append("name", name)
            .append("kind", kind)
            .append("superTypeDeclarations", superTypeDeclarations)
            .append("innerFields", innerFields)
            .append("innerMethods", innerMethods)
            .append("innerConstructors", innerConstructors)
            .append("innerRecords", innerRecords)
            .toString()
    }

    @DoNotPersist
    override val eogStarters: List<Node>
        get() {
            val list = mutableListOf<Node>()

            list += innerFields
            list += innerMethods
            list += innerConstructors

            return list
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RecordDeclaration) return false
        return super.equals(other) &&
            kind == other.kind &&
            innerFields == other.innerFields &&
            propertyEqualsList(innerFieldEdges, other.innerFieldEdges) &&
            innerMethods == other.innerMethods &&
            propertyEqualsList(innerMethodEdges, other.innerMethodEdges) &&
            innerConstructors == other.innerConstructors &&
            propertyEqualsList(innerConstructorEdges, other.innerConstructorEdges) &&
            innerRecords == other.innerRecords &&
            propertyEqualsList(innerRecordEdges, other.innerRecordEdges) &&
            superClasses == other.superClasses &&
            implementedInterfaces == other.implementedInterfaces &&
            superTypeDeclarations == other.superTypeDeclarations
    }

    override fun hashCode() = super.hashCode() // TODO: Which fields can be safely added?

    override fun addDeclaration(declaration: Declaration) {
        when (declaration) {
            is ConstructorDeclaration -> addIfNotContains(innerConstructorEdges, declaration)
            is MethodDeclaration -> addIfNotContains(innerMethodEdges, declaration)
            is FieldDeclaration -> addIfNotContains(innerFieldEdges, declaration)
            is RecordDeclaration -> addIfNotContains(innerRecordEdges, declaration)
            is TemplateDeclaration -> addIfNotContains(templateEdges, declaration)
        }
    }

    /**
     * Returns a type represented by this record.
     *
     * @return the type
     */
    fun toType(): ObjectType {
        val type = objectType(name)
        if (type is ObjectType) {
            // As a shortcut, directly set the record declaration. This will be otherwise done
            // later by a pass, but for some frontends we need this immediately, so we set
            // this here.
            type.recordDeclaration = this
            type.superTypes.addAll(this.superTypes)
            return type
        } else {
            throw TranslationException("Cannot create type for $this, as it is not an ObjectType")
        }
    }

    override val declaredType: Type
        get() = toType()

    override val secondaryTypes: List<Type>
        get() = superTypes
}
